processor/src/lib.rs

```diff
@@ -277,21 +277,25 @@ where
             MastNode::Call(node) => self.execute_call_node(node, program),
             MastNode::Dyn => self.execute_dyn_node(program),
             MastNode::External(external_node) => {
-                let mast_forest =
-                    self.host.borrow().get_mast_forest(&external_node.digest()).ok_or_else(
-                        || ExecutionError::MastForestNotFound {
-                            root_digest: external_node.digest(),
-                        },
-                    )?;
+                let node_digest = external_node.digest();
+                let mast_forest = self
+                    .host
+                    .borrow()
+                    .get_mast_forest(&node_digest)
+                    .ok_or(ExecutionError::MastForestNotFound { root_digest: node_digest })?;
 
                 // We limit the parts of the program that can be called externally to procedure
                 // roots, even though MAST doesn't have that restriction.
-                let root_id = mast_forest.find_procedure_root(external_node.digest()).ok_or(
-                    ExecutionError::MalformedMastForestInHost {
-                        root_digest: external_node.digest(),
-                    },
+                let root_id = mast_forest.find_procedure_root(node_digest).ok_or(
+                    ExecutionError::MalformedMastForestInHost { root_digest: node_digest },
                 )?;
 
+                // if the node that we got by looking up an external reference is also an External
+                // node, we are about to enter into an infinite loop - so, return an error
+                if mast_forest[root_id].is_external() {
+                    return Err(ExecutionError::CircularExternalNode(node_digest));
+                }
```

Comment:
`call` would not resolve to an `External` node (it would resolve to a `Call` node) - but if we substitute `call` with `exec`, that should be an error.

Basically, if we have `exec.f1` where `f1` is is `External(f1_digest)` and we look it up and get `External(f1_digest)` again, we are in an infinite loop (if we get `External(some_other_digest)` during lookup - it is a different error: we asked for a node with digest `x` but got `y`).